doc/diskseekd.1
.TH diskseekd 1 "03Mar05" fdutils-5.5
.SH Name
diskseek, diskseekd - disk seek daemon; simulates Messy Dos' drive cleaning effect
'\" t
.de TQ
.br
.ns
.TP \\$1
..

.tr \(is'
.tr \(if`
.tr \(pd"

.SH Note
This manpage has been automatically generated from fdutils's texinfo
documentation.  However, this process is only approximative, and some
items, such as cross-references, footnotes and indices are lost in this
translation process.  Indeed, these items have no appropriate
representation in the manpage format.  Moreover, only the items specific
to each command have been translated, and the general information about
fdutils has been dropped in the manpage version.  Thus I strongly advise
you to use the original texinfo doc.
.TP
* \ \ 
To generate a printable copy from the texinfo doc, run the following
commands:
 
.nf
.ft 3
.in +0.3i
    ./configure; make dvi; dvips fdutils.dvi
.fi
.in -0.3i
.ft R
 
\&\fR
.TP
* \ \ 
To generate a HTML copy,  run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make html
.fi
.in -0.3i
.ft R
 
\&\fRA pre-made HTML can be found at:
\&\fR\&\f(CW\(ifhttp://www.tux.org/pub/knaff/fdutils\(is\fR
.TP
* \ \ 
To generate an info copy (browsable using emacs' info mode), run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make info
.fi
.in -0.3i
.ft R
 
\&\fR
.PP
The texinfo doc looks most pretty when printed or as HTML.  Indeed, in
the info version certain examples are difficult to read due to the
quoting conventions used in info.
.SH Description
.PP
Several people have noticed that Linux has a bad tendency of killing
floppy drives. These failures remained completely mysterious, until
somebody noticed that they were due to huge layers of dust accumulating
in the floppy drives. This cannot happen under Messy Dos, because this
excuse for an operating system is so unstable that it crashes roughly
every 20 minutes (actually less if you are running Windows).  When
rebooting, the BIOS seeks the drive, and by doing this, it shakes the
dust out of the drive mechanism. \fR\&\f(CWdiskseekd\fR simulates this effect
by seeking the drive periodically.  If it is called as \fR\&\f(CWdiskseek\fR,
the drive is sought only once.
.PP
.SH Options
.PP
The syntax for \fR\&\f(CWdiskseekd\fR is as follows:
 
.nf
.ft 3
.in +0.3i
\&\fR\&\f(CWdiskseekd [\fR\&\f(CW\-d \fIdrive\fR\&\f(CW] [\fR\&\f(CW\-i \fIinterval\fR\&\f(CW] [\fR\&\f(CW\-p \fIpidfile\fR\&\f(CW]
.fi
.in -0.3i
.ft R
 
\&\fR
.TP
\&\fR\&\f(CW\-d\ \fIdrive\fR\&\f(CW\fR\ 
Selects the drive to seek.  By default, drive 0 (\fR\&\f(CW\(if/dev/fd0\(is\fR) is sought.
.TP
\&\fR\&\f(CW\-i\ \fIinterval\fR\&\f(CW\fR\ 
Selects the cleaning interval, in seconds.  If the interval is 0, a
single seek is done. This is useful when calling diskseek from a
crontab.  The default is 1000 seconds (about 16 minutes) for
\&\fR\&\f(CWdiskseekd\fR and 0 for \fR\&\f(CWdiskseek\fR.
.TP
\&\fR\&\f(CW\-p\ \fIpidfile\fR\&\f(CW\fR\ 
Stores the process id of the diskseekd daemon into \fIpidfile\fR instead
of the default \fR\&\f(CW\(if/var/run/diskseekd.pid\(is\fR.
.PP
.SH Bugs
.TP
1.\ 
Other aspects of Messy Dos' flakiness are not simulated.
.TP
2.\ 
This section lacks a few smileys.
.SH See Also
Fdutils' texinfo doc


Changelog
020203	Fix fdmount and superformat bugs
020203	Create /dev/fd0 before testing for existence of /dev/fd0
011002	Fix floppycontrol.c printf conditional on FD_DISK_CHANGED, so that
	it works with gcc3 which apparently considers printf as a macro...
000612	Explicitly include sysmacros.h, apparently this is needed in
	certain environments
000603	Released 5.4
000527	Fixed a couple of buffer overrun exploits and other cleanup. Made
	fdutils compilable with Linux 2.3/2.4 kernels. Documentation fixes
970104	Fixed a few bugs listed in the Debian buglist posted to
	linux.debian.bugs.dis
	1.  Lessen dependancy of superformat on mtools by not failing if
	mformat cannot be called: low-level verify will still be done, but
	mbadblocks will be skipped
	2.  Fixed a few instances where the return value of getopt was
	misdeclared as a char instead of an integer
	3.  Fixed device names in fd.4
970411	Added error check to capacity measurement
	Corrected documentation Makefile bug
970629	Added vfat support to fdmount
970609	Fixed bug where a skew was erroneously applied to a 2m track
	Fixed typo in printfdprm
970604	Added vfat option for fdmount
970516	5.2 Released
970516	Fixed sunsite location in doc.
970430	Fixed typoes in doc about autodetection.
970428	Fixed manpage installation
970427	5.1 Released.  These repeated complaints about yylineno not
	defined start getting annoying ;-)
970426	More Makefile fixes.  Added note to doc about Alpha site
970422	Various Makefile fixes
970421	Lex yylineno fixes
970419	5.0 Released


src/diskseekd.c
#include <sys/types.h>
#ifdef HAVE_SYS_SYSMACROS_H
# include <sys/sysmacros.h>
#endif
#include <sys/time.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <linux/fd.h>
#include <linux/fdreg.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <linux/major.h>
#include "enh_options.h"
#include <sys/signal.h>
#include <errno.h>

int eioctl(int fd, int command,void * param, char *emsg)
{
  int r;
  if ((r=ioctl(fd,command,param))<0 )
    {
      perror(emsg);
      exit(1);
    }
  return r;
}

#ifndef PIDFILE
#define PIDFILE NULL
#endif

void dummy(int dummy)
{}

int main(int argc, char **argv)
{
	int mask=0;
	int fd=-2;
	int interval=1000;
	int ch;
	struct timeval timval;
	struct floppy_raw_cmd raw_cmd;
	struct stat buf;
	int drive;
	char *drive_name="/dev/fd0";
	char *pidfilename=PIDFILE;
	char *tmp_s;
	FILE *pidfile;
	int pid;

	struct enh_options optable[] = {
	{ 'd', "drive", 1, EO_TYPE_STRING, 0, 0,
		  (void *) &drive_name,
		  "drive to be seeked"},

	{ 'i', "interval", 1, EO_TYPE_LONG, 0, 0, 
		  (void *) &interval,
		  "set polling interval (in tenth of seconds)" },

	{ 'p', "pidfile", 1, EO_TYPE_STRING, 0, 0,
		  (void *) &pidfilename,
		  "stores the daemon process id into named file"},

	{ '\0', 0 }
	};

	tmp_s = strrchr(argv[0],'/');
	if(!tmp_s)
		tmp_s = argv[0];
	else
		tmp_s++;
	if ( !strcmp(tmp_s, "diskseek" ))
		interval = 0;

	while((ch=getopt_enh(argc, argv, optable, 
			     0, &mask, "") ) != EOF ){
		if ( ch== '?' ){
			fprintf(stderr,"exiting\n");
			exit(1);
		}
		printf("unhandled option %d\n", ch);
		exit(1);
	}

	if(interval){
		switch((pid = fork())){
		case 0: /* child */
			break;
		case -1: /* error */
			perror("fork");
			exit(1);
		default: /* parent */
			if(!pidfilename)
				pidfilename = getenv("DISKSEEKDPIDFILE");
			if(pidfilename){				
				pidfile = fopen(pidfilename, "w");
				fprintf(pidfile, "%d\n", pid);
				fclose(pidfile);
			}
			exit(0);
		}
	}

	while(1){
		signal(SIGHUP, dummy);
		/* open the device */
		fd=open(drive_name, O_ACCMODE | O_EXCL);

		if (fd < 0 ){
			fprintf(stderr,"%s\n", drive_name);
			perror("open");
			exit(1);
		}

		if (fstat (fd, &buf) < 0) {
			perror("fstat");
			exit(1);
		}
		if (major(buf.st_rdev) != FLOPPY_MAJOR) {
			fprintf(stderr,"Not a floppy drive\n");
			exit(1);
		}

		drive = minor( buf.st_rdev );
		drive = (drive & 3) + ((drive & 0x80) >> 5);

		/* reset the fdc, if needed */
		eioctl(fd, FDRESET, FD_RESET_IF_NEEDED, "reset");

		/* then recalibrate the disk */
		raw_cmd.flags = FD_RAW_INTR;
		raw_cmd.cmd_count = 2;
		raw_cmd.cmd[0] = FD_RECALIBRATE;
		raw_cmd.cmd[1] = drive;
		eioctl(fd, FDRAWCMD, (void *)&raw_cmd, "recalibrate");

		raw_cmd.flags = FD_RAW_INTR;
		raw_cmd.cmd_count = 3;
		raw_cmd.cmd[0] = FD_SEEK;
		raw_cmd.cmd[1] = drive & 3;
		raw_cmd.cmd[2] = 80;
		eioctl(fd, FDRAWCMD, (void *)&raw_cmd, "blind seek");
			
		/* then recalibrate the disk */
		raw_cmd.flags = FD_RAW_INTR;
		raw_cmd.cmd_count = 2;
		raw_cmd.cmd[0] = FD_RECALIBRATE;
		raw_cmd.cmd[1] = drive;
		eioctl(fd, FDRAWCMD, (void *)&raw_cmd, "recalibrate");
		close(fd);

		if(!interval)
			exit(0);
		timval.tv_sec = interval;
		timval.tv_usec = 0;
		if (select(0, 0, 0, 0, &timval) < 0 && errno != EINTR){
			perror("select");
			exit(1);
		}
	}
    return 0;
}
